File: typescript/src/agents/comprehendFilterAgent.ts
import { Agent, AgentOptions } from "./agent";
import { ConversationMessage, ParticipantRole } from "../types";
import { Logger } from "../utils/logger";
import {
  ComprehendClient,
  DetectSentimentCommand,
  DetectPiiEntitiesCommand,
  DetectToxicContentCommand,
  DetectSentimentCommandOutput,
  DetectPiiEntitiesCommandOutput,
  DetectToxicContentCommandOutput,
  LanguageCode
} from "@aws-sdk/client-comprehend";

// Interface for toxic content labels returned by Comprehend
interface ToxicContent {
  Name: "GRAPHIC" | "HARASSMENT_OR_ABUSE" | "HATE_SPEECH" | "INSULT" | "PROFANITY" | "SEXUAL" | "VIOLENCE_OR_THREAT";
  Score: number;
}

// Interface for toxic labels result structure
interface ToxicLabels {
  Labels: ToxicContent[];
  Toxicity: number;
}

// Type definition for custom check functions
type CheckFunction = (input: string) => Promise<string | null>;

// Extended options for ComprehendContentFilterAgent
export interface ComprehendFilterAgentOptions extends AgentOptions {
    region?: string;
    enableSentimentCheck?: boolean;
    enablePiiCheck?: boolean;
    enableToxicityCheck?: boolean;
    sentimentThreshold?: number;
    toxicityThreshold?: number;
    allowPii?: boolean;
    languageCode?: LanguageCode;
}

/**
 * ComprehendContentFilterAgent class
 *
 * This agent uses Amazon Comprehend to analyze and filter content based on
 * sentiment, PII, and toxicity. It can be configured to enable/disable specific
 * checks and allows for the addition of custom checks.
 */
export class ComprehendFilterAgent extends Agent {
  private comprehendClient: ComprehendClient;
  private customChecks: CheckFunction[] = [];

  private enableSentimentCheck: boolean;
  private enablePiiCheck: boolean;
  private enableToxicityCheck: boolean;
  private sentimentThreshold: number;
  private toxicityThreshold: number;
  private allowPii: boolean;
  private languageCode: LanguageCode;

  /**
   * Constructor for ComprehendContentFilterAgent
   * @param options - Configuration options for the agent
   */
  constructor(options: ComprehendFilterAgentOptions) {
    super(options);

    this.comprehendClient = options.region
      ? new ComprehendClient({ region: options.region })
      : new ComprehendClient();

    // Set default configuration using fields from options
    this.enableSentimentCheck = options.enableSentimentCheck ?? true;
    this.enablePiiCheck = options.enablePiiCheck ?? true;
    this.enableToxicityCheck = options.enableToxicityCheck ?? true;
    this.sentimentThreshold = options.sentimentThreshold ?? 0.7;
    this.toxicityThreshold = options.toxicityThreshold ?? 0.7;
    this.allowPii = options.allowPii ?? false;
    this.languageCode = this.validateLanguageCode(options.languageCode) ?? 'en';

    // Ensure at least one check is enabled
    if (!this.enableSentimentCheck &&
        !this.enablePiiCheck &&
        !this.enableToxicityCheck) {
      this.enableToxicityCheck = true;
    }
  }

  /**
   * Processes a user request by sending it to the Amazon Bedrock agent for processing.
   * @param inputText - The user input as a string.
   * @param userId - The ID of the user sending the request.
   * @param sessionId - The ID of the session associated with the conversation.
   * @param chatHistory - An array of Message objects representing the conversation history.
   * @param additionalParams - Optional additional parameters as key-value pairs.
   * @returns A Promise that resolves to a Message object containing the agent's response.
   */
  /* eslint-disable @typescript-eslint/no-unused-vars */
  async processRequest(
    inputText: string,
    userId: string,
    sessionId: string,
    chatHistory: ConversationMessage[],
    additionalParams?: Record<string, string>
  ): Promise<ConversationMessage> {
    try {
      const issues: string[] = [];

      // Run all checks in parallel
      const [sentimentResult, piiResult, toxicityResult] = await Promise.all([
        this.enableSentimentCheck ? this.detectSentiment(inputText) : null,
        this.enablePiiCheck ? this.detectPiiEntities(inputText) : null,
        this.enableToxicityCheck ? this.detectToxicContent(inputText) : null
      ]);

      // Process results
      if (this.enableSentimentCheck && sentimentResult) {
        const sentimentIssue = this.checkSentiment(sentimentResult);
        if (sentimentIssue) issues.push(sentimentIssue);
      }

      if (this.enablePiiCheck && piiResult) {
        const piiIssue = this.checkPii(piiResult);
        if (piiIssue) issues.push(piiIssue);
      }

      if (this.enableToxicityCheck && toxicityResult) {
        const toxicityIssue = this.checkToxicity(toxicityResult);
        if (toxicityIssue) issues.push(toxicityIssue);
      }

      // Run custom checks
      for (const check of this.customChecks) {
        const customIssue = await check(inputText);
        if (customIssue) issues.push(customIssue);
      }

      if (issues.length > 0) {
        Logger.logger.warn(`Content filter issues detected: ${issues.join('; ')}`);
        return null;  // Return null to indicate content should not be processed further
      }

      // If no issues, return the original input as a ConversationMessage
      return {
        role: ParticipantRole.ASSISTANT,
        content: [{ text: inputText }]
      };

    } catch (error) {
      Logger.logger.error("Error in ComprehendContentFilterAgent:", error);
      throw error;
    }
  }

  /**
   * Add a custom check function to the agent
   * @param check - A function that takes a string input and returns a Promise<string | null>
   */
  addCustomCheck(check: CheckFunction) {
    this.customChecks.push(check);
  }

  /**
   * Check sentiment of the input text
   * @param result - Result from Comprehend's sentiment detection
   * @returns A string describing the issue if sentiment is negative, null otherwise
   */
  private checkSentiment(result: DetectSentimentCommandOutput): string | null {
    if (result.Sentiment === 'NEGATIVE' &&
        result.SentimentScore?.Negative > this.sentimentThreshold) {
      return `Negative sentiment detected (${result.SentimentScore.Negative.toFixed(2)})`;
    }
    return null;
  }

  /**
   * Check for PII in the input text
   * @param result - Result from Comprehend's PII detection
   * @returns A string describing the issue if PII is detected, null otherwise
   */
  private checkPii(result: DetectPiiEntitiesCommandOutput): string | null {
    if (!this.allowPii && result.Entities && result.Entities.length > 0) {
      return `PII detected: ${result.Entities.map(e => e.Type).join(', ')}`;
    }
    return null;
  }

  /**
   * Check for toxic content in the input text
   * @param result - Result from Comprehend's toxic content detection
   * @returns A string describing the issue if toxic content is detected, null otherwise
   */
  private checkToxicity(result: DetectToxicContentCommandOutput): string | null {
    const toxicLabels = this.getToxicLabels(result);
    if (toxicLabels.length > 0) {
      return `Toxic content detected: ${toxicLabels.join(', ')}`;
    }
    return null;
  }

  /**
   * Detect sentiment using Amazon Comprehend
   * @param text - Input text to analyze
   */
  private async detectSentiment(text: string) {
    const command = new DetectSentimentCommand({
      Text: text,
      LanguageCode: this.languageCode
    });
    return this.comprehendClient.send(command);
  }

  /**
   * Detect PII entities using Amazon Comprehend
   * @param text - Input text to analyze
   */
  private async detectPiiEntities(text: string) {
    const command = new DetectPiiEntitiesCommand({
      Text: text,
      LanguageCode: this.languageCode
    });
    return this.comprehendClient.send(command);
  }

  /**
   * Detect toxic content using Amazon Comprehend
   * @param text - Input text to analyze
   */
  private async detectToxicContent(text: string) {
    const command = new DetectToxicContentCommand({
      TextSegments: [{ Text: text }],
      LanguageCode: this.languageCode
    });
    return this.comprehendClient.send(command);
  }

  /**
   * Extract toxic labels from the Comprehend response
   * @param toxicityResult - Result from Comprehend's toxic content detection
   * @returns Array of toxic label names that exceed the threshold
   */
  private getToxicLabels(toxicityResult: DetectToxicContentCommandOutput): string[] {
    const toxicLabels: string[] = [];

    if (toxicityResult.ResultList && Array.isArray(toxicityResult.ResultList)) {
      toxicityResult.ResultList.forEach((result: ToxicLabels) => {
        if (result.Labels && Array.isArray(result.Labels)) {
          result.Labels.forEach((label: ToxicContent) => {
            if (label.Score > this.toxicityThreshold) {
              toxicLabels.push(label.Name);
            }
          });
        }
      });
    }

    return toxicLabels;
  }

  /**
   * Set the language code for Comprehend operations
   * @param languageCode - The ISO 639-1 language code
   */
  setLanguageCode(languageCode: LanguageCode): void {
    const validatedLanguageCode = this.validateLanguageCode(languageCode);
    if (validatedLanguageCode) {
      this.languageCode = validatedLanguageCode;
    } else {
      throw new Error(`Invalid language code: ${languageCode}`);
    }
  }

  /**
   * Validate the provided language code
   * @param languageCode - The language code to validate
   * @returns The validated LanguageCode or undefined if invalid
   */
  private validateLanguageCode(languageCode: LanguageCode | undefined): LanguageCode | undefined {
    if (!languageCode) return undefined;

    const validLanguageCodes: LanguageCode[] = [
      'en', 'es', 'fr', 'de', 'it', 'pt', 'ar', 'hi', 'ja', 'ko', 'zh', 'zh-TW'
    ];

    return validLanguageCodes.includes(languageCode) ? languageCode : undefined;
  }
}

File: python/src/multi_agent_orchestrator/agents/comprehend_filter_agent.py
from typing import List, Dict, Union, Optional, Callable, Any
from multi_agent_orchestrator.types import ConversationMessage, ParticipantRole
from multi_agent_orchestrator.utils.logger import Logger
from .agent import Agent, AgentOptions
import boto3
from botocore.config import Config

# Type alias for CheckFunction
CheckFunction = Callable[[str], str]

class ComprehendFilterAgentOptions(AgentOptions):
    def __init__(self,
                 enable_sentiment_check: bool = True,
                 enable_pii_check: bool = True,
                 enable_toxicity_check: bool = True,
                 sentiment_threshold: float = 0.7,
                 toxicity_threshold: float = 0.7,
                 allow_pii: bool = False,
                 language_code: str = 'en',
                 **kwargs):
        super().__init__(**kwargs)
        self.enable_sentiment_check = enable_sentiment_check
        self.enable_pii_check = enable_pii_check
        self.enable_toxicity_check = enable_toxicity_check
        self.sentiment_threshold = sentiment_threshold
        self.toxicity_threshold = toxicity_threshold
        self.allow_pii = allow_pii
        self.language_code = language_code

class ComprehendFilterAgent(Agent):
    def __init__(self, options: ComprehendFilterAgentOptions):
        super().__init__(options)

        config = Config(region_name=options.region) if options.region else None
        self.comprehend_client = boto3.client('comprehend', config=config)

        self.custom_checks: List[CheckFunction] = []

        self.enable_sentiment_check = options.enable_sentiment_check
        self.enable_pii_check = options.enable_pii_check
        self.enable_toxicity_check = options.enable_toxicity_check
        self.sentiment_threshold = options.sentiment_threshold
        self.toxicity_threshold = options.toxicity_threshold
        self.allow_pii = options.allow_pii
        self.language_code = self.validate_language_code(options.language_code) or 'en'

        # Ensure at least one check is enabled
        if not any([self.enable_sentiment_check, self.enable_pii_check, self.enable_toxicity_check]):
            self.enable_toxicity_check = True

    async def process_request(self,
                              input_text: str,
                              user_id: str,
                              session_id: str,
                              chat_history: List[ConversationMessage],
                              additional_params: Optional[Dict[str, str]] = None) -> Optional[ConversationMessage]:
        try:
            issues: List[str] = []

            # Run all checks
            sentiment_result = self.detect_sentiment(input_text) if self.enable_sentiment_check else None
            pii_result = self.detect_pii_entities(input_text) if self.enable_pii_check else None
            toxicity_result = self.detect_toxic_content(input_text) if self.enable_toxicity_check else None

            # Process results
            if self.enable_sentiment_check and sentiment_result:
                sentiment_issue = self.check_sentiment(sentiment_result)
                if sentiment_issue:
                    issues.append(sentiment_issue)

            if self.enable_pii_check and pii_result:
                pii_issue = self.check_pii(pii_result)
                if pii_issue:
                    issues.append(pii_issue)

            if self.enable_toxicity_check and toxicity_result:
                toxicity_issue = self.check_toxicity(toxicity_result)
                if toxicity_issue:
                    issues.append(toxicity_issue)

            # Run custom checks
            for check in self.custom_checks:
                custom_issue = await check(input_text)
                if custom_issue:
                    issues.append(custom_issue)

            if issues:
                Logger.warn(f"Content filter issues detected: {'; '.join(issues)}")
                return None  # Return None to indicate content should not be processed further

            # If no issues, return the original input as a ConversationMessage
            return ConversationMessage(
                role=ParticipantRole.ASSISTANT.value,
                content=[{"text": input_text}]
            )

        except Exception as error:
            Logger.error(f"Error in ComprehendContentFilterAgent:{str(error)}")
            raise error

    def add_custom_check(self, check: CheckFunction):
        self.custom_checks.append(check)

    def check_sentiment(self, result: Dict[str, Any]) -> Optional[str]:
        if result['Sentiment'] == 'NEGATIVE' and result['SentimentScore']['Negative'] > self.sentiment_threshold:
            return f"Negative sentiment detected ({result['SentimentScore']['Negative']:.2f})"
        return None

    def check_pii(self, result: Dict[str, Any]) -> Optional[str]:
        if not self.allow_pii and result.get('Entities'):
            return f"PII detected: {', '.join(e['Type'] for e in result['Entities'])}"
        return None

    def check_toxicity(self, result: Dict[str, Any]) -> Optional[str]:
        toxic_labels = self.get_toxic_labels(result)
        if toxic_labels:
            return f"Toxic content detected: {', '.join(toxic_labels)}"
        return None

    def detect_sentiment(self, text: str) -> Dict[str, Any]:
        return self.comprehend_client.detect_sentiment(
            Text=text,
            LanguageCode=self.language_code
        )

    def detect_pii_entities(self, text: str) -> Dict[str, Any]:
        return self.comprehend_client.detect_pii_entities(
            Text=text,
            LanguageCode=self.language_code
        )

    def detect_toxic_content(self, text: str) -> Dict[str, Any]:
        return self.comprehend_client.detect_toxic_content(
            TextSegments=[{"Text": text}],
            LanguageCode=self.language_code
        )

    def get_toxic_labels(self, toxicity_result: Dict[str, Any]) -> List[str]:
        toxic_labels = []
        for result in toxicity_result.get('ResultList', []):
            for label in result.get('Labels', []):
                if label['Score'] > self.toxicity_threshold:
                    toxic_labels.append(label['Name'])
        return toxic_labels

    def set_language_code(self, language_code: str):
        validated_language_code = self.validate_language_code(language_code)
        if validated_language_code:
            self.language_code = validated_language_code
        else:
            raise ValueError(f"Invalid language code: {language_code}")

    @staticmethod
    def validate_language_code(language_code: Optional[str]) -> Optional[str]:
        if not language_code:
            return None

        valid_language_codes = ['en', 'es', 'fr', 'de', 'it', 'pt', 'ar', 'hi', 'ja', 'ko', 'zh', 'zh-TW']
        return language_code if language_code in valid_language_codes else None